File: API/_archive/GraphQL/python/comprehensive_production_graphql.py
import strawberry
from fastapi import FastAPI, Depends, HTTPException, Request
from strawberry.fastapi import GraphQLRouter
from typing import List, Optional

# --- Mock Database ---
users_db = {
    "u1": {"id": "u1", "name": "Alice", "email": "alice@example.com"},
    "u2": {"id": "u2", "name": "Bob", "email": "bob@example.com"},
}

posts_db = {
    "p1": {"id": "p1", "title": "GraphQL with Strawberry", "author_id": "u1"},
    "p2": {"id": "p2", "title": "FastAPI integration", "author_id": "u2"},
}

# --- Context & Authentication Dependency ---
def get_current_user(request: Request):
    """
    FastAPI dependency to extract the token from headers.
    This injects the user context into Strawberry GraphQL resolvers.
    """
    token = request.headers.get("Authorization")
    if token == "Bearer admin-token":
        return "u1"
    return None

async def get_context(current_user: Optional[str] = Depends(get_current_user)):
    return {"user_id": current_user}

# --- GraphQL Types ---
@strawberry.type
class User:
    id: str
    name: str
    email: str

@strawberry.type
class Post:
    id: str
    title: str
    author_id: strawberry.Private[str] # Private field, not exposed in API

    # Nested Resolver: Resolves the author relationship dynamically
    @strawberry.field
    def author(self) -> User:
        user_data = users_db.get(self.author_id)
        if not user_data:
            raise Exception("Author not found")
        return User(**user_data)

# --- Queries (Reads) ---
@strawberry.type
class Query:
    
    # Simple query returning a list
    @strawberry.field
    def all_posts(self) -> List[Post]:
        return [Post(**data) for data in posts_db.values()]

    # Query with arguments
    @strawberry.field
    def post_by_id(self, id: str) -> Optional[Post]:
        data = posts_db.get(id)
        if data:
            return Post(**data)
        return None

# --- Mutations (Writes) ---
@strawberry.type
class Mutation:
    
    @strawberry.mutation
    def create_post(self, info: strawberry.Info, title: str) -> Post:
        # Access the context (Injected by FastAPI dependencies)
        user_id = info.context.get("user_id")
        
        if not user_id:
            raise Exception("Unauthorized: Missing or invalid token")
            
        new_id = f"p{len(posts_db) + 1}"
        new_post_data = {"id": new_id, "title": title, "author_id": user_id}
        posts_db[new_id] = new_post_data
        
        return Post(**new_post_data)

# --- Schema & FastAPI Integration ---
schema = strawberry.Schema(query=Query, mutation=Mutation)

graphql_app = GraphQLRouter(
    schema,
    context_getter=get_context # Binds FastAPI auth to Strawberry context
)

app = FastAPI(title="Production GraphQL API")
app.include_router(graphql_app, prefix="/graphql")

if __name__ == "__main__":
    import uvicorn
    print("Run with: uvicorn comprehensive_production_graphql:app --reload")
    # uvicorn.run(app, host="0.0.0.0", port=8000)

File: API/_archive/GraphQL/golang/comprehensive_production_graphql.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"

	"github.com/graphql-go/graphql"
)

// --- Models & Mock DB ---
type User struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Email string `json:"email"`
}

type Post struct {
	ID       string `json:"id"`
	Title    string `json:"title"`
	AuthorID string `json:"author_id"`
}

var usersDB = map[string]User{
	"u1": {ID: "u1", Name: "Alice", Email: "alice@example.com"},
	"u2": {ID: "u2", Name: "Bob", Email: "bob@example.com"},
}

var postsDB = map[string]Post{
	"p1": {ID: "p1", Title: "GraphQL is awesome", AuthorID: "u1"},
	"p2": {ID: "p2", Title: "REST vs GraphQL", AuthorID: "u2"},
}

// --- Middleware (Authentication) ---
func AuthMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		token := r.Header.Get("Authorization")
		// In production, validate JWT here
		userID := ""
		if token == "Bearer admin-token" {
			userID = "u1" // Mock identified user
		}

		// Inject user into context
		ctx := context.WithValue(r.Context(), "userID", userID)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// --- GraphQL Types ---
var userType = graphql.NewObject(graphql.ObjectConfig{
	Name: "User",
	Fields: graphql.Fields{
		"id":    &graphql.Field{Type: graphql.String},
		"name":  &graphql.Field{Type: graphql.String},
		"email": &graphql.Field{Type: graphql.String},
	},
})

var postType = graphql.NewObject(graphql.ObjectConfig{
	Name: "Post",
	Fields: graphql.Fields{
		"id":    &graphql.Field{Type: graphql.String},
		"title": &graphql.Field{Type: graphql.String},
		// Advanced: Resolving nested fields
		"author": &graphql.Field{
			Type: userType,
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				if post, ok := p.Source.(Post); ok {
					return usersDB[post.AuthorID], nil
				}
				return nil, nil
			},
		},
	},
})

// --- Queries ---
var queryType = graphql.NewObject(graphql.ObjectConfig{
	Name: "Query",
	Fields: graphql.Fields{
		// 1. Get a specific post by ID (Arguments)
		"post": &graphql.Field{
			Type: postType,
			Args: graphql.FieldConfigArgument{
				"id": &graphql.ArgumentConfig{Type: graphql.NewNonNull(graphql.String)},
			},
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				id := p.Args["id"].(string)
				post, exists := postsDB[id]
				if !exists {
					return nil, fmt.Errorf("post not found")
				}
				return post, nil
			},
		},
		// 2. Get all posts
		"posts": &graphql.Field{
			Type: graphql.NewList(postType),
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				var result []Post
				for _, v := range postsDB {
					result = append(result, v)
				}
				return result, nil
			},
		},
	},
})

// --- Mutations ---
var mutationType = graphql.NewObject(graphql.ObjectConfig{
	Name: "Mutation",
	Fields: graphql.Fields{
		"createPost": &graphql.Field{
			Type: postType,
			Args: graphql.FieldConfigArgument{
				"title": &graphql.ArgumentConfig{Type: graphql.NewNonNull(graphql.String)},
			},
			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
				// Access Context (Authentication check)
				ctx := p.Context
				userID, ok := ctx.Value("userID").(string)
				if !ok || userID == "" {
					return nil, fmt.Errorf("unauthorized: must provide valid Bearer token")
				}

				newPost := Post{
					ID:       fmt.Sprintf("p%d", len(postsDB)+1),
					Title:    p.Args["title"].(string),
					AuthorID: userID,
				}
				postsDB[newPost.ID] = newPost
				return newPost, nil
			},
		},
	},
})

func main() {
	// Compile Schema
	schemaConfig := graphql.SchemaConfig{Query: queryType, Mutation: mutationType}
	schema, err := graphql.NewSchema(schemaConfig)
	if err != nil {
		log.Fatalf("failed to create new schema, error: %v", err)
	}

	// HTTP Handler
	graphqlHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			Query     string                 `json:"query"`
			Operation string                 `json:"operationName"`
			Variables map[string]interface{} `json:"variables"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			http.Error(w, "Error parsing JSON request body", 400)
			return
		}

		// Execute GraphQL request with the HTTP Context (containing auth data)
		result := graphql.Do(graphql.Params{
			Schema:         schema,
			RequestString:  req.Query,
			VariableValues: req.Variables,
			OperationName:  req.Operation,
			Context:        r.Context(),
		})

		json.NewEncoder(w).Encode(result)
	})

	// Mount middleware and start server
	http.Handle("/graphql", AuthMiddleware(graphqlHandler))
	log.Println("GraphQL server running on :8080/graphql")
	http.ListenAndServe(":8080", nil)
}
